Reg Admin: scroll-to section and sections summary

kr-matthews opened the pull request:
Some ideas. The main point is to allow scrolling to a section. Then I thought having information about the section would be good (and might alleviate the need to scroll), so I added that as a separate thing. I could combine them by making the summary table rows clickable and removing the "scroll to" menu option.

~I also combined the two email options into a dropdown to save space (on mobile), but I suspect there may be more complaints about needing 2 clicks (though I think we only got 1 complaint about moving registrations?). I'm thinking about reverting it.~

~I'm not very happy about the count in "Move ${count} to" since the export and email options should also have it for consistency. Having it in them all 3 doesn't make sense, so having it once, outside the buttons somewhere would make sense. But I can't come up with a good place.~

Thoughts? ~Any ideas to fix the Popup table on mobile (see last video)?~

It's a bit annoying that we aren't consistently using keys across the front-end code, e.g. waiting vs waitlist, accepted vs approved. I can 'fix' it, but I think that would require changing a few translation keys, which isn't ideal. _Edit: I kind of fixed it, via a helper function to translate status key to translation key._

:warning: **Images and screenshots here are out-of-date, see the latest comment at the bottom.** :warning:

![image](https://github.com/user-attachments/assets/228f5498-bc90-4c27-9b13-16f63041a6ce)
![image](https://github.com/user-attachments/assets/308a09b5-8b91-4d7c-87c3-5575d7403bb7)

Competition without competitor limit:
![image](https://github.com/user-attachments/assets/96d4744a-60e0-4e91-9a53-4ab5a28f5c62)
![image](https://github.com/user-attachments/assets/95e71767-2e0f-4b4f-b6fe-53d9c08a0239)

Competition with competitor limit:
![image](https://github.com/user-attachments/assets/938615a1-e29d-4ba4-a671-6df97151b935)
![image](https://github.com/user-attachments/assets/90db0d02-04d3-40a9-ab42-a6f7bb34b169)

A small offset added to the scroll-to functionality, otherwise the header is hidden under the buttons...
[Screencast from 2025-03-06 10:30:57 PM.webm](https://github.com/user-attachments/assets/bcdec4df-80a2-489f-922a-07fa143c1b30)
...although that happens on mobile anyway :\ and the popup summary table doesn't quite work.
[Screencast from 2025-03-06 10:35:03 PM.webm](https://github.com/user-attachments/assets/7acde978-6e79-4cb0-a175-3b517cb290fb)

kr-matthews committed:
Combine email actions into dropdown

kr-matthews committed:
Add refs to table sections

kr-matthews committed:
Add dropdown to scroll to sections

kr-matthews committed:
And summary table on hover

kr-matthews committed:
Position popup bottom left

gregorbg commented on app/webpacker/components/Persons/NewPersonForm/NewPersonForm.js:
```diff
@@ -140,7 +137,7 @@ function NewPersonForm({
             dob,
             gender,
-            country_id: countryIdForIso2(countryIso2),
+            countryId: countryIdForIso2(countryIso2),
```
Just curious, why did you change this from `snake_case` to `camelCase`? The backend usually expects snake case and I don't see any relevant changes to the backend in this PR...

kr-matthews replied:
Uh this whole file looks like it might just be a mishandled rebase conflict. I didn't check carefully before pushing, since it's just a draft and I'm more looking for feedback on the UI rather than on the code, at this point.

gregorbg commented on app/webpacker/components/RegistrationsV2/RegistrationAdministration/RegistrationActions.jsx:
```diff
@@ -342,7 +429,50 @@ export default function RegistrationActions({
+  withMaximums,
+}) {
+  return (
+    <Table basic="very">
```
All of the below table components can use dot notation on the original `Table` import, for example `Table.Header` or `Table.HeaderCell` (and others). This saves you a lot of import statements at the top.

gregorbg commented on app/webpacker/components/RegistrationsV2/RegistrationAdministration/RegistrationAdministrationList.jsx:
```diff
@@ -525,12 +541,14 @@ export default function RegistrationAdministrationList({ competitionInfo }) {
-      <Icon name={icon} color={color} />
-      {`${title} (${inParens})`}
-    </Header>
+    <span ref={sectionRef} style={{ scrollMarginTop: '7em' }}>
```
Does the `Header` component itself not support `ref` natively? If not, I think SemUI has a `Ref` component that you can use to pass refs down within their own framework

gregorbg replied:
(we do actually use that component in a couple of places already, so just search for `<Ref` throughout our codebase and you will see)

gregorbg commented on app/webpacker/components/RegistrationsV2/RegistrationAdministration/RegistrationAdministrationList.jsx:
```diff
@@ -68,6 +68,15 @@ export default function RegistrationAdministrationList({ competitionInfo }) {
+  const approvedRef = useRef();
+  const cancelledRef = useRef();
+  const rejectedRef = useRef();
+  const tableRefs = useMemo(() => ({
+    pendingRef, waitlistRef, approvedRef, cancelledRef, rejectedRef,
+  }), []);
```
I like the concise notation, but I am a bit afraid it would make the code "vulnerable" to refactors. What about using the actual `competition_status` enum values (like `"pending"`, `"approved"`, etc) as keys?
That would make it a bit more verbose here but more standardized and easier to maintain in the future

kr-matthews committed:
Fix NewPersonForm miss-rebase

kr-matthews committed:
Use dot notation for table imports

kr-matthews committed:
Use Ref instead of span

kr-matthews committed:
Use standard keys for scroll-to refs

kr-matthews committed:
Allow scrolling to non-competing list when visible

kr-matthews committed:
Introduce registrationStatusKeys and de-duplicate

kr-matthews committed:
Condense switch statements

kr-matthews committed:
Deduplicate move dropdown code

kr-matthews committed:
Use status helpers instead of consts directly

kr-matthews committed:
Replace registrationStatusTranslationKeys

with getStatusTranslationKey

kr-matthews committed:
registrationStatusKeys arg to handle nonCompeting

kr-matthews committed:
No labels on buttons, use only icons & tooltip

kr-matthews committed:
Remove unused DropdownLink

kr-matthews committed:
Improve export_csv translation

kr-matthews committed:
Fix move_to.one plural

kr-matthews commented:
Not 100% done (tooltips are sometimes not visible, the table doesn't work well on mobile, email tooltip missing text, etc.) but would like feedback on the direction of the UI in particular. Thinking of adding total money to the summary table too, but don't want it to get too crowded.

[Screencast from 2025-09-07 08:29:32 PM.webm](https://github.com/user-attachments/assets/1480eeac-0f86-45bc-a47a-ee596fbd7b1e)


[Screencast from 2025-09-07 08:23:36 PM.webm](https://github.com/user-attachments/assets/692df1de-2a2d-41bd-9fcd-cfeebc6dc97e)

kr-matthews committed:
Switch summary table to a grid

& use icons for the statuses

kr-matthews committed:
Switch from data-tooltip to Popup

and adjust email text

kr-matthews committed:
Move action buttons down, space for popups

kr-matthews committed:
Adjust scroll offset to match button offset

kr-matthews commented:
⚠️ Edit: these videos are out-of-date. ⚠️

Only issue is that the notifications are hidden behind the buttons. I'll fix that next time.

(Oops I forgot to check the box to record the cursor position... Hopefully it's obvious.)

[10980.webm](https://github.com/user-attachments/assets/e1b47223-9110-4e33-9a53-2c32f0d9903a)

[noncompeting.webm](https://github.com/user-attachments/assets/f1cd12b2-a2b0-44e3-b5d2-7c7f10d25cab)

The table didn't work on mobile because it was automatically becoming a stackable table. So I used a grid instead. The grid is by default wide (about twice as wide as the mobile screen width), so I added an explicit width. If there's a better way, let me know.

The popups need to be on top, otherwise they interfere with the dropdowns. To have them on top, I had to move the buttons down.

I could also add another "$" button which would show money totals per registration status (when payments are enabled).

kr-matthews committed:
Show non-competing option in summary when non-zero

kr-matthews committed:
Move sticky registration message above buttons

kr-matthews commented:
Message moved up:

<img width="556" height="594" alt="image" src="https://github.com/user-attachments/assets/f5fa3512-d499-4dda-b086-02d7835db2d9" />

kr-matthews committed:
Merge branch 'main' into feature/reg-admin-scroll-to-table

kr-matthews committed:
Merge branch 'main' into feature/reg-admin-scroll-to-table

kr-matthews committed:
Merge branch 'main' into feature/reg-admin-scroll-to-table

gregorbg reviewed (commented):
I'm fine with the code but I need some degree of user-testing on this. My personal experience with the day-to-day usage of the Reg Admin page is limited.

@dunkOnIT Can you play around with this to give a third-party opinion on whether these changes all make sense from a UX perspective?

dunkOnIT commented:
As mentioned in DM's to Kevin, I really like the changes - especially the infobox/summary box showing all the registrations! 

Notes/suggestions from me: 
- [ ] Do you think we could combine the infobox pictured above and the scroll-to functionality? The idea being that if the cursor is over the modal that comes up, it stays up - and the icon can be clickd on to scroll to a particular section
- [ ] Add explainer text for "Approved": "Approved registrations are confirmed to be able to participate in the competition in their registered events." 
- [ ] Infobox table sizing in Desktop view looks a little squished - I think either less vertical padding in the cells, or more width in the columns, would alleviate this 

@kr-matthews feedback finalized - apologies, forgot to remove my "still in progress" disclaimer at the top of this comment

kr-matthews committed:
Merge branch 'main' into feature/reg-admin-scroll-to-table

kr-matthews committed:
Add counts as hints on scroll-to menu items

kr-matthews committed:
Remove summary table and info icon

kr-matthews committed:
Add header text for accepted table

kr-matthews committed:
Rename pluralized keys

kr-matthews committed:
Spell "e.g.," correctly

kr-matthews committed:
Fix menu item text overlap

kr-matthews commented:
I got rid of the info icon & popup table, and just put the counts as a hint/description on the move-to options. This is simpler than a table. It doesn't include the competitor limit any more, but that was a bit awkward anyway as it only applied to at most one section.

<img width="760" height="531" alt="image" src="https://github.com/user-attachments/assets/f364fd11-711b-4b28-b58b-64e7b242ee5a" />

https://github.com/user-attachments/assets/b9d359ba-62cb-4fd1-b303-792b28f2d2a7

https://github.com/user-attachments/assets/23df93bd-276a-434d-b00b-1e3d20d79de4

The only downside is that I had to manually specify a (min) width for the menu items. As mentioned [in the documentation](https://semantic-ui.com/modules/dropdown.html#:~:text=Using%20a%20description,Filter%20Tags), this is necessary (when the item is in a menu). I added enough width to handle 4-digit registration totals (e.g., "1000" or "1000 of 1000"). It's not ideal though.

<img width="734" height="371" alt="image" src="https://github.com/user-attachments/assets/34016a35-b697-4dce-a18c-ca3ae7b292aa" />

I also added the approved-registrations text that Duncan requested, and renamed a few translation keys which I had pluralized (apparently not renaming them would cause issues for translators), and fixed "e.g.," usages.

dunkOnIT commented:
Amazing, thanks Kevin! Happy with it from a functionality perspective :+1:

kr-matthews committed:
Merge branch 'main' into feature/reg-admin-scroll-to-table

kr-matthews commented:
I don't understand why the tests keep failing

FinnIckler commented:
Tests are failing because you changed the spelling in the en.yml 
`(eg., Stripe) to (e.g., Stripe)`. Let me push a fix

FinnIckler committed:
fix tests
